### ISSUE TITLE
Some learnings from Cordova templates

### ISSUE THREAD

**Issue by axemclion** [mkonicek/AppTemplateFeedback#1]
Sometime ago, we introduced the concept of templates in Cordova and I am hoping that some of the learning from it can be applied here 

1. Custom templates, specified using `--template` is super useful. A lot of companies have their own starter projects, so allowing people to publish and use their own templates is great
2. Looks like the current proposal only picks up templates from npm. A lot of companies may want templates in private github repos. Can the `--template` also take a URL (http://, file:// and git://) so that custom templates published and shared withing organizations ?
3. One common thing thing about templates is app secrets and keys. Whenever someone publishes a template with a backend call, they would want to let the template user change database URLs, keys, etc. React Native already has the link command. Is there a way for templates to call the link command to configure these things ?
4. Many templates would also want to use native plugins. Not sure how the custom template will support that. Searching for the right plugin is also a little hard - having a template with "recommended" plugins will be super useful.

**Comment by mkonicek** [comment]
All of those points are great, thank you!

2. Good idea! Will add that.
3. Good point. This will be a bit involved but hopefully we can come up with a simple way to do string replacements. Will look into that.
4. A template can specify dependencies using a file called [dependencies.json](https://github.com/facebook/react-native/blob/master/local-cli/templates/HelloNavigation/dependencies.json). Currently 'react-native init --template' does 'npm install --save' (or 'yarn add' if you have yarn) for each of those dependencies. I can call 'react-native link' after all those deps are installed.

Regarding recommended modules - it depends what functionality an app needs and I think we shouldn't add lots of modules (e.g. video player, maps) by default even if the app doesn't need them. This is because extra modules increase app size. But having a list of recommended modules would be useful. Right now this can be used as that list more or less: https://js.coach/react-native/generator-react-native-ignite?sort=popular.

**Comment by axemclion** [comment]
Regarding native modules, I agree that it should not be a part of the default template you create - but developers would want to add their own native modules. My suggestion was to have a way developers can add native modules to the templates they create.

**Comment by ericvicenti** [comment]
Personally I think it is too early to add a URL input to RN init.

We already have:

`react-native init`
`create-react-app`

And soon we'll have `create-react-native-app` and maybe `create-react-component` or something.

Ultimately I'd like to have a more flexible generator system, but I think we should try to get on the same page as the other generators.

**Comment by axemclion** [comment]
Agreed that a generator is probably a good place to scaffold complex apps. 

In that case, would it now make sense to let `react-native init` create a simple, blank app and move the functionality discussed here to a generator, or a create-react-native-app instead ?

**Comment by mkonicek** [comment]
> Regarding native modules, I agree that it should not be a part of the default template you create - but developers would want to add their own native modules. My suggestion was to have a way developers can add native modules to the templates they create. I think a dependencies.json would be easy.

I agree. This will be supported. The template can specify any dependencies (including native ones) in its dependencies.json.

**Comment by mkonicek** [comment]
> But then, how would this be different from Ignite, or other starter generators ?

Indeed it's pretty similar. The difference between the RN templates and Ignite is the RN templates can be published by anyone, and the ones included by default will be less opinionated than Ignite. Wrote a bit about this here: https://github.com/mkonicek/AppTemplateFeedback#how-is-this-different-from-ignite

Note that Ignite can build on top of the RN templates if they decide to:

`react-native init MyApp --template ignite`

**Comment by geirman** [comment]
Great discussion! I'm trying to figure out what the syntax is for installing from git:// and http:// as mentioned above by axemclion. Can someone throw me a bone please?

**Comment by DeveloperZH** [comment]
if i want to change android project  build.gradle, hao to do it?
